File: 01-what-is-cpp/1-time-cpp-vs-python/count_permutations.py
#!python3

import itertools
import time

def count_permutations(N:int)->int:
	count=0
	for p in itertools.permutations(range(1,N+1)):
		count += 1
	return count

if __name__=="__main__":
	for N in range(1,13):
		print ("Permutations of 1..{}:".format(N), flush=True)
		start = time.time()
		count = count_permutations(N) 
		end = time.time()
		duration_in_seconds = end-start
		print("  {} permutations calculated in {} seconds".format(count, duration_in_seconds), flush=True)

File: 01-what-is-cpp/1-time-cpp-vs-python/count_permutations.cpp
#include "output_containers.hpp"
#include <algorithm>
#include <numeric>
using namespace std;

int count_permutations(int N) {
	vector<int> v(N);
	iota(v.begin(), v.end(), 1); // fill the vector with 1,...,N
	int count=0;
	do {
		++count;
	} while ( next_permutation(v.begin(),v.end()) );
	return count;
}

int main() {
	for (int N=1; N<14; ++N) {
		cout << "Permutations of 1.." << N << ":" << endl;
		auto begin = time(NULL);
		auto count = count_permutations(N);
		auto end = time(NULL);
		auto duration = end-begin;
		cout << "  " << count << " permutations calculated in " << duration << " seconds" << endl;
	}
	return 0;
}
